Sources/NIO/EventLoopFuture.swift

```diff
@@ -957,6 +957,43 @@ extension EventLoopFuture {
     }
 }
 
+extension EventLoopFuture {
+    /// Returns a new `EventLoopFuture` that succeeds when all of the provided `EventLoopFuture`s complete.
+    /// The new `EventLoopFuture` will contain an array of results, maintaining ordering for each of the `EventLoopFuture`s.
+    ///
+    /// The returned `EventLoopFuture` always succeeds, regardless of any failures from the waiting futures.
+    ///
+    /// If it is desired to flatten them into a single `EventLoopFuture` that fails on the first `EventLoopFuture` failure,
+    /// use one of the `reduce` methods instead.
+    /// - Parameter futures: An array of homogenous `EventLoopFuture`s to gather results from.
+    /// - Returns: A new `EventLoopFuture` with all the results of the provided futures.
+    public static func whenAllComplete<InputValue>(_ futures: [EventLoopFuture<InputValue>],
+                                                   eventLoop: EventLoop) -> EventLoopFuture<[Result<InputValue, Error>]> {
+        return eventLoop.makeSucceededFuture(result: [])
+            .flatMap { _ in
+                let promise = eventLoop.makePromise(of: [Result<InputValue, Error>].self)
+
+                var remainingCount = futures.count
+                var results: [Result<InputValue, Error>?] = .init(repeating: nil, count: futures.count)
+
+                for (index, future) in futures.enumerated() {
+                    future
+                        .hopTo(eventLoop: eventLoop)
+                        ._whenComplete {
+                            results[index] = future.value!
+                            remainingCount -= 1
+
+                            guard remainingCount == 0 else { return CallbackList() }
+
+                            return promise._setValue(value: .success(results.compactMap({ $0 })))
+                        }
```

Comment:
This block uses a load of `EventLoopFuture` SPI: let's not. Per #184 I'd like to reduce the amount of internal SPI used in `EventLoopFuture` rather than increase it, and as we can write all of this without SPI, I think we should.

That means replacing `._whenComplete` with `whenComplete`, removing `future.value!` (as it'll no longer be necessary), and replacing `promise._setValue` with `promise.succeed`.

Reply:
Thanks for the context - I was going off of what I saw in the file already. I'll make the updates to not use SPI.